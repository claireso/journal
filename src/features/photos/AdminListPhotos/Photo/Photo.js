import React, { useRef, useCallback } from 'react'
import PropTypes from 'prop-types'

import * as S from './Photo.styles'

import { PhotoTypes } from '@types'

import useInView from '@hooks/useInView'

import AnimatedImage from '@components/AnimatedImage'
import { IconPencil, IconDelete } from '@components/Icons'
import { ButtonIcon } from '@components/Buttons'

const Photo = ({ id, color, source, title, description, onEdit, onDelete }) => {
  const dom = useRef(null)
  const inView = useInView(dom)

  const handleEdit = useCallback(() => onEdit(id), [id, onEdit])
  const handleDelete = useCallback(() => onDelete(id), [id, onDelete])

  return (
    <S.PhotoWrapper ref={dom} data-testid="photo">
      <S.PhotoPicture style={{ color: color || 'transparent' }}>
        {inView && <AnimatedImage src={source} />}
      </S.PhotoPicture>
      <S.PhotoInner>
        {title && <S.PhotoTitle dangerouslySetInnerHTML={{ __html: title }} />}
        {description && <S.PhotoDescription>{description}</S.PhotoDescription>}
      </S.PhotoInner>
      <S.PhotoTools>
        <ButtonIcon data-testid="button-edit" onClick={handleEdit} title="Edit">
          <IconPencil />
        </ButtonIcon>
        <ButtonIcon data-testid="button-delete" onClick={handleDelete} title="Delete">
          <IconDelete />
        </ButtonIcon>
      </S.PhotoTools>
    </S.PhotoWrapper>
  )
}

Photo.propTypes = {
  ...PhotoTypes,
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
}

export default Photo
