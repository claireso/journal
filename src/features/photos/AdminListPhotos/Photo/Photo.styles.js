import { styled } from '@theme'

export const PhotoWrapper = styled('li', {
  alignItems: 'center',
  display: 'grid',
  gridTemplateColumns: '8rem auto 12rem',
  gridColumnGap: '$4',
  p: '$5',
  transition: '$background',
  '&:hover': {
    background: '$secondary100'
  },
  '& + &': {
    borderTop: '1px solid $secondary200'
  }
})

export const PhotoPicture = styled('div', {
  height: '8rem',
  img: {
    height: '100%',
    objectFit: 'contain',
    width: '100%',
    filter: `
      drop-shadow(0 -6px 0 currentColor)
      drop-shadow(0 6px 0 currentColor)
      drop-shadow(-6px 0 0 currentColor)
      drop-shadow(6px 0 0 currentColor)
    `
  }
})

export const PhotoInner = styled('div', {
  flex: 1
})

export const PhotoTitle = styled('h2', {
  fontSize: '$4',
  m: '$1 0 $3',
  fontWeight: '$semiBold'
})

export const PhotoDescription = styled('p', {
  color: '$gray500',
  fontSize: '$2'
})

export const PhotoTools = styled('p', {
  placeSelf: 'center flex-end',
  transition: '$opacity',
  button: {
    opacity: 0.55,
    transition: '$opacity',
    '&:hover': {
      opacity: 1
    }
  }
})
