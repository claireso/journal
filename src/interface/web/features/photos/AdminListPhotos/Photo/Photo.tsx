import React from 'react'

import * as cls from './styles.css'

import { PhotoDto } from '@dto'
import { formatDate } from '@utils/date'

import Media from './Media'
import Actions from './Actions'

import Text from '@web/components/Text'

type PhotoProps = PhotoDto

const Photo = ({ id, color, media, title, description, created_at }: PhotoProps) => {
  return (
    <div className={cls.wrapper}>
      <Media {...media} color={color}>
        <Actions photoId={id} />
      </Media>
      <div className={cls.content}>
        <div className={cls.contentInner}>
          {title && (
            <Text as="span" size="sm">
              <span dangerouslySetInnerHTML={{ __html: title }} />
            </Text>
          )}
          {description && (
            <Text as="span" size="sm" color="neutral">
              <span dangerouslySetInnerHTML={{ __html: description }} />
            </Text>
          )}
        </div>
        <Text as="span" size="xxs" color="neutral" className={cls.date}>
          {formatDate(created_at)}
        </Text>
      </div>
    </div>
  )
}

export default Photo
