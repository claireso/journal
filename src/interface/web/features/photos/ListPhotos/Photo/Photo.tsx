'use client'

import React, { useRef } from 'react'
import type { PhotoDto } from '@dto'

import AnimatedImage from '@web/components/AnimatedImage'
import Text from '@web/components/Text'
import useInView from '@web/hooks/useInView'
import { tokens } from '@web/theme/core/tokens.css'

import * as cls from './styles.css'

interface PhotoProps extends PhotoDto {
  row: number
}

const Photo = ({ title, description, media, position, color, row }: PhotoProps) => {
  const dom = useRef<HTMLElement>(null!)
  const isInView = useInView(dom)

  return (
    <figure
      className={cls.figure({
        landscape: !media.portrait && !media.square,
        portrait: media.portrait,
        square: media.square,
        position: position,
        highlight: !!color
      })}
      ref={dom}
      style={{
        gridRowStart: row
      }}
    >
      <div
        className={cls.pictureWrapper({
          portrait: media.portrait,
          square: media.square,
          highlight: !!color
        })}
        style={{
          color: color || tokens.colors.neutral['2extralight']
        }}
      >
        {isInView && <AnimatedImage className={cls.picture} src={media.source} />}
      </div>
      <figcaption className={cls.figcaption}>
        {title && (
          <Text as="span" size="sm">
            <span dangerouslySetInnerHTML={{ __html: title }} />
          </Text>
        )}
        {description && (
          <Text as="span" size="xs" color="neutral" italic>
            <span dangerouslySetInnerHTML={{ __html: description }} />
          </Text>
        )}
      </figcaption>
    </figure>
  )
}

export default Photo
