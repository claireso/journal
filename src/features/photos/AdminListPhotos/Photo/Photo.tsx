import React, { useRef, useCallback } from 'react'

import * as S from './Photo.styles'

import { EnhancedPhoto } from '@models'
import useInView from '@hooks/useInView'

import AnimatedImage from '@components/AnimatedImage'
import { IconPencil, IconDelete } from '@components/Icons'
import { ButtonIcon } from '@components/Buttons'

interface PhotoProps extends EnhancedPhoto {
  onEdit: (id: number) => void
  onDelete: (id: number) => void
}

const Photo = ({ id, color, source, title, description, onEdit, onDelete }: PhotoProps) => {
  const dom = useRef(null)
  const inView = useInView(dom)

  const handleEdit = useCallback(() => onEdit(id), [id, onEdit])
  const handleDelete = useCallback(() => onDelete(id), [id, onDelete])

  return (
    <S.PhotoWrapper ref={dom} data-testid="photo">
      <S.PhotoPicture style={{ color: color || 'transparent' }}>
        {inView && <AnimatedImage src={source} />}
      </S.PhotoPicture>
      <S.PhotoInner>
        {title && <S.PhotoTitle dangerouslySetInnerHTML={{ __html: title }} />}
        {description && <S.PhotoDescription>{description}</S.PhotoDescription>}
      </S.PhotoInner>
      <S.PhotoTools>
        <ButtonIcon data-testid="button-edit" onClick={handleEdit} title="Edit">
          <IconPencil />
        </ButtonIcon>
        <ButtonIcon data-testid="button-delete" onClick={handleDelete} title="Delete">
          <IconDelete />
        </ButtonIcon>
      </S.PhotoTools>
    </S.PhotoWrapper>
  )
}

export default Photo
