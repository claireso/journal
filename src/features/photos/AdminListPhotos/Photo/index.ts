import Photo from './Photo'

export default Photo
