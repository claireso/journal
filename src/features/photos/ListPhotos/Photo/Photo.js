import React, { useRef } from 'react'
import PropTypes from 'prop-types'

import { PhotoTypes } from '@types'

import useInView from '@hooks/useInView'

import * as S from './Photo.styles'

const Photo = ({ title, description, source, portrait, position, square, color, row }) => {
  const dom = useRef(null)
  const isInView = useInView(dom)

  return (
    <S.Figure
      ref={dom}
      data-testid="photo"
      portrait={portrait}
      square={square}
      position={position}
      withColor={!!color}
      css={{
        gridRowStart: row
      }}
    >
      <S.PictureWrapper
        css={{
          color: color || '$secondary200',
          '--aspect-ratio': portrait ? 2 / 3 : square ? 1 / 1 : 3 / 2
        }}
      >
        {isInView && <S.Picture src={source} alt="" />}
      </S.PictureWrapper>
      <S.Title>
        <span dangerouslySetInnerHTML={{ __html: title }} />

        {description && <S.Description>{description}</S.Description>}
      </S.Title>
    </S.Figure>
  )
}

Photo.propTypes = {
  ...PhotoTypes,
  row: PropTypes.number.isRequired
}

export default Photo
