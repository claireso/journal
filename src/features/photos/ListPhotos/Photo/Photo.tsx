import React, { useRef } from 'react'

import useInView from '@hooks/useInView'

import * as S from './Photo.styles'

interface PhotoProps extends Photo {
  row: number
}

const Photo = ({ title, description, source, portrait, position, square, color, row }: PhotoProps) => {
  const dom = useRef(null)
  const isInView = useInView(dom)

  return (
    <S.Figure
      ref={dom}
      data-testid="photo"
      portrait={portrait}
      square={square}
      // @ts-ignore
      position={position}
      withColor={!!color}
      css={{
        gridRowStart: row
      }}
    >
      <S.PictureWrapper
        css={{
          color: color || '$secondary200',
          '--aspect-ratio': portrait ? 2 / 3 : square ? 1 / 1 : 3 / 2
        }}
      >
        {isInView && <S.Picture src={source} />}
      </S.PictureWrapper>
      <S.Title>
        {title && <span dangerouslySetInnerHTML={{ __html: title }} />}
        {description && <S.Description>{description}</S.Description>}
      </S.Title>
    </S.Figure>
  )
}

export default Photo
