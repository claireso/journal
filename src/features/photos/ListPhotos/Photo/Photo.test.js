import { render } from '@testing-library/react'

import Photo from './index'

describe('<Photo />', () => {
  const renderPhoto = (props = {}) => render(<Photo {...global.__PHOTO__} row={0} {...props} />)

  test('should render photo (landscape center)', () => {
    const { container } = renderPhoto()

    expect(container).toMatchSnapshot()
  })

  test('should render photo (landscape left)', () => {
    const { container } = renderPhoto({ position: 'left' })

    expect(container).toMatchSnapshot()
  })

  test('should render photo (landscape right)', () => {
    const { container } = renderPhoto({ position: 'right' })

    expect(container).toMatchSnapshot()
  })

  test('should render photo (portrait center)', () => {
    const { container } = renderPhoto({ portrait: true })

    expect(container).toMatchSnapshot()
  })

  test('should render photo (square center)', () => {
    const { container } = renderPhoto({ square: true })

    expect(container).toMatchSnapshot()
  })
})
