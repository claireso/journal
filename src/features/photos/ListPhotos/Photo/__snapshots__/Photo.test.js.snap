// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Photo /> should render photo (landscape center) 1`] = `
<DocumentFragment>
  <figure
    class="c-jmBjsY c-jmBjsY-iiCDtyl-css"
    data-testid="photo"
  >
    <div
      class="c-igxWur c-igxWur-ibEeMWi-css"
    />
    <figcaption
      class="c-iMVRpC"
    >
      <span>
        Single photography
      </span>
      <span
        class="c-hTMuHN"
      >
        Janvier 2019
      </span>
    </figcaption>
  </figure>
</DocumentFragment>
`;

exports[`<Photo /> should render photo (landscape left) 1`] = `
<DocumentFragment>
  <figure
    class="c-jmBjsY c-jmBjsY-iiCDtyl-css"
    data-testid="photo"
  >
    <div
      class="c-igxWur c-igxWur-ibEeMWi-css"
    />
    <figcaption
      class="c-iMVRpC"
    >
      <span>
        Single photography
      </span>
      <span
        class="c-hTMuHN"
      >
        Janvier 2019
      </span>
    </figcaption>
  </figure>
</DocumentFragment>
`;

exports[`<Photo /> should render photo (landscape right) 1`] = `
<DocumentFragment>
  <figure
    class="c-jmBjsY c-jmBjsY-iiCDtyl-css"
    data-testid="photo"
  >
    <div
      class="c-igxWur c-igxWur-ibEeMWi-css"
    />
    <figcaption
      class="c-iMVRpC"
    >
      <span>
        Single photography
      </span>
      <span
        class="c-hTMuHN"
      >
        Janvier 2019
      </span>
    </figcaption>
  </figure>
</DocumentFragment>
`;

exports[`<Photo /> should render photo (portrait center) 1`] = `
<DocumentFragment>
  <figure
    class="c-jmBjsY c-jmBjsY-gxKxyj-portrait-true c-jmBjsY-bHfBnH-cv c-jmBjsY-iiCDtyl-css"
    data-testid="photo"
  >
    <div
      class="c-igxWur c-igxWur-idhNoZG-css"
    />
    <figcaption
      class="c-iMVRpC"
    >
      <span>
        Single photography
      </span>
      <span
        class="c-hTMuHN"
      >
        Janvier 2019
      </span>
    </figcaption>
  </figure>
</DocumentFragment>
`;

exports[`<Photo /> should render photo (square center) 1`] = `
<DocumentFragment>
  <figure
    class="c-jmBjsY c-jmBjsY-gzEPDe-square-true c-jmBjsY-iiCDtyl-css"
    data-testid="photo"
  >
    <div
      class="c-igxWur c-igxWur-ievjdZN-css"
    />
    <figcaption
      class="c-iMVRpC"
    >
      <span>
        Single photography
      </span>
      <span
        class="c-hTMuHN"
      >
        Janvier 2019
      </span>
    </figcaption>
  </figure>
</DocumentFragment>
`;
